packages/playwright-core/src/cli/docker.ts

```diff
@@ -0,0 +1,389 @@
+    process.exit(1);
+  }
+
+  let info = await containerInfo();
+  if (info)
+    return info;
+
+  await launchContainer({
+    image: pwImage,
+    name: VRT_CONTAINER_NAME,
+    autoRemove: true,
+    ports: [5400, 7900],
+  });
+
+  // Wait for the service to become available.
+  const startTime = Date.now();
+  const timeouts = [0, 100, 100, 200, 500, 1000];
+  do {
+    await new Promise(x => setTimeout(x, timeouts.shift() ?? 1000));
+    info = await containerInfo();
+  } while (!info && Date.now() < startTime + 60000);
+
+  if (!info)
+    throw new Error('Failed to launch docker container!');
+  return info;
+}
+
+export async function stopContainer() {
+  const containerId = await findRunningDockerContainerId();
+  if (!containerId) {
+    console.log(`Container is not running.`);
+    return undefined;
+  }
+  await Promise.all([
+    // Make sure to wait for the container to be removed.
+    postJSON(`/containers/${containerId}/wait?condition=removed`),
+    postJSON(`/containers/${containerId}/kill`),
+  ]);
+}
+
+export async function ensureDockerEngineIsRunningOrDie() {
+  try {
+    await callDockerAPI('get', '/info');
+  } catch (e) {
+    console.error(utils.wrapInASCIIBox([
+      `Docker is not running!`,
+      `Please install if necessary and launch docker first:`,
+      ``,
+      `    https://docs.docker.com/get-docker`,
+      ``,
+    ].join('\n'), 1));
+    process.exit(1);
+  }
+}
+
+async function findDockerImage(imageName: string): Promise<DockerImage|undefined> {
+  const images: DockerImage[] | null = await getJSON('/images/json');
+  return images ? images.find(image => image.RepoTags?.includes(imageName)) : undefined;
+}
+
+interface Container {
+  ImageID: string;
+  State: string;
+  Names: [string];
+  Id: string;
+}
+
+async function findRunningDockerContainerId(): Promise<string|undefined> {
+  const containers: (Container[]|undefined) = await getJSON('/containers/json');
+  if (!containers)
+    return undefined;
+  // 1. Try findind a container with our name. This happens if the container is launched
+  // automatically with `npx playwright docker start`.
+  let container = containers.find((container: Container) => container.Names.some(name => name.includes(VRT_CONTAINER_NAME)));
+  // 2. Alternatively, the container might be launched manually with a direct docker command.
+  // In this case, we should look for a container with a proper base image.
+  if (!container) {
+    const dockerImage = await findDockerImage(VRT_IMAGE_NAME);
+    container = dockerImage ? containers.find((container: Container) => container.ImageID === dockerImage.Id) : undefined;
+  }
+  return container?.State === 'running' ? container.Id : undefined;
+}
+
+interface ContainerOptions {
+  image: DockerImage;
+  autoRemove: boolean;
+  command?: string[];
+  ports?: Number[];
+  name?: string;
+}
+
+async function launchContainer(options: ContainerOptions): Promise<string> {
+  const ExposedPorts: any = {};
```

Comment:
Follow-up ideas:
- `docker stop` and `docker delete-image` should stop all versions of pw image;
- bind to available ports on the host instead of 5400 and 7900;